src/Compiler/Utilities/Caches.fs

```diff
@@ -164,58 +171,100 @@ type Cache<'Key, 'Value when 'Key: not null and 'Key: equality> internal (totalC
     let evicted = Event<_>()
     let evictionFailed = Event<_>()
 
-    let cts = new CancellationTokenSource()
+    // Track disposal state
+    let mutable disposed = false
```

Comment:
Should this be volatile or a ref cell + do some sort of interlocked exchange/volatile readwrite? Can something be racing in the dispose? Do we care?

Reply:
Yes good catch. Currently the only cache we create is a global singleton, but in the long run this should be rock solid, if we aim to use it in more places and with shorter lifetimes.